libopenage/curve/array.h

```diff
@@ -0,0 +1,259 @@
+
+	Array(const Array &) = delete;
+	Array &operator=(const Array &) = delete;
+
+
+	/**
+	 * Get the last element with elem->time <= time from
+	 * the keyfram container at a given index
+	 */
+	T at(const time::time_t &t, const size_t index) const;
+
+
+	/**
+	 * Get an array of the last elements with elem->time <= time from
+	 * all keyfram containers contained within this array curve
+	 */
+	std::array<T, Size> get(const time::time_t &t) const;
+
+	// Get the amount of KeyframeContainers in array curve
+	consteval size_t size() const;
+
+	/**
+	 * Get the last element and associated time which has elem->time <= time from
+	 * the keyfram container at a given index
+	 */
+	std::pair<time::time_t, T> frame(const time::time_t &t, const size_t index) const;
+
+
+	/**
+	 * Get the first element and associated time which has elem->time >= time from
+	 * the keyfram container at a given index
+	 */
+	std::pair<time::time_t, T> next_frame(const time::time_t &t, const size_t index) const;
+
+
+	/**
+	 * Insert a new keyframe value at time t at index
+	 */
+	void set_insert(const time::time_t &t, const size_t index, T value);
+
+
+	/**
+	 * Insert a new keyframe value at time t; delete all keyframes after time t at index
+	 */
+	void set_last(const time::time_t &t, const size_t index, T value);
+
+
+	/**
+	 * Insert a new keyframe value at time t at index i; remove all other keyframes with time t at index i
+	 */
+	void set_replace(const time::time_t &t, const size_t index, T value);
+
+
+	/**
+	 * Copy keyframes from another container to this container.
+	 *
+	 * Replaces all keyframes beginning at t >= start with keyframes from \p other.
+	 *
+	 * @param other Curve that keyframes are copied from.
+	 * @param start Start time at which keyframes are replaced (default = -INF).
+	 *              Using the default value replaces ALL keyframes of \p this with
+	 *              the keyframes of \p other.
+	 */
+	void sync(const Array<T, Size> &other, const time::time_t &start);
+
+
+	/**
+	 * Get the identifier of this curve.
+	 *
+	 * @return Identifier.
+	 */
+	size_t id() const override {
+		return this->_id;
+	}
+
+	/**
+	 * Get the human-readable identifier of this curve.
+	 *
+	 * @return Human-readable identifier.
+	 */
+	std::string idstr() const override {
+		if (this->_idstr.size() == 0) {
+			return std::to_string(this->id());
+		}
+		return this->_idstr;
+	}
+
+
+	// get a copy to the KeyframeContainer at index
+	KeyframeContainer<T> operator[](size_t index) const {
+		return this->container.at(index);
+	}
+
+	// Array::Iterator is used to iterate over KeyframeContainers contained in a curve at a given time.
+	class Iterator {
+	public:
+		Iterator(Array<T, Size> *curve, const time::time_t &time = time::TIME_MAX, size_t offset = 0) :
+			curve(curve), time(time), offset(offset) {};
+
+		// returns a copy of the keyframe at the current offset and time
+		const T operator*() {
+			return this->curve->frame(this->time, this->offset).second;
+		}
```

Comment:
1. This method doesn't return a keyframe, it returns `T`
2. If it only returns `T`, you should use `at()` instead of `frame()` to retrieve the value.